api: blocks/{round}/logs endpoint

joe-p opened the pull request:
## Summary

This is a new endpoint to get all of the logs from all appcalls in a given block. This endpoint was created specifically due to the encoding issue with the current `/blocks/{round}` endpoint (#5807). 

## Test Plan

Tests cover the following scenarios:

* log in outer transaction
* log from inner transaction
* log from newly created application
* logs from multiple outer app calls

jannotti commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,54 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
 	return ctx.JSON(http.StatusOK, response)
 }
 
+func getLogsFromTxns(txns []transactions.SignedTxnWithAD, blockLogs *[]model.BlockLog, outerTxnID string) {
+
```
You generally want to just return the slice, rather than pass a slice object in by reference. (Slices in Go are small, they pretty much _are_ a reference.)

jannotti commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,54 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
+func getLogsFromTxns(txns []transactions.SignedTxnWithAD, blockLogs *[]model.BlockLog, outerTxnID string) {
+
+	for _, txn := range txns {
+		byteLogs := [][]byte{}
```
We like to make the slice the right size when we know it, so there's no growth cost.
```suggestion
		byteLogs := make([][]byte, len(txn.EvalDelta.Logs))
```

jannotti replied:
This point can be ignored now, as I see you actually _don't_ know the eventual size, if you intend to recur for inners of inners.

joe-p replied:
Since we are mapping the array of logs to app index and transaction ID, this array won't be increased from inners... although I suppose a single outer app could call the same app multiple times. The current behavior is to just create a new struct for each call, even if it's to an app that's already been called, but I'm open to a different approach.

jannotti replied:
Oh, you're collecting all of a given app's logs into one struct, not one per top-level txn?  I missed that somehow.

That seems like it will remove a lot of information about what happened.

jannotti replied:
But I also don't follow what you're saying. If you don't descend the `txn.EvalDelta.Inners[..].EvalDelta.Inners[...].EvalDelta.Inners[....]` you're never going to get all the logs.

joe-p replied:
It does descend down through inners: https://github.com/joe-p/go-algorand/blob/6d4b23f72649f7a2fd42176ca420a424872dedcd/daemon/algod/api/server/v2/handlers.go#L749. 

I realized though I'm not actually ever getting the logs from the outers

jannotti replied:
Ah, ok, I saw that line and somehow forgot where it was.

jannotti commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,54 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
+
+	for _, txn := range txns {
+		byteLogs := [][]byte{}
+		for _, log := range txn.EvalDelta.Logs {
+			byteLogs = append(byteLogs, []byte(log))
```
And then you would:
```suggestion
		for i, log := range txn.EvalDelta.Logs {
			byteLogs[i] = []byte(log)
```

jannotti replied:
More importantly, you're going to have to go digging through that `EvalDelta` to find all the logs, if you want them for inners of inners too.

jannotti commented on daemon/algod/api/algod.oas2.json:
```diff
@@ -4441,6 +4441,33 @@
         }
       }
+    },
+    "BlockLog": {
+      "description": "An array of logs for a given application ID and outer transaction ID",
```
I think you want a slightly different schema for your response.  I think you have:
```
[ { "txid": BC12433, "application-index": 345, "items": [ a bunch of strings] },
  { "txid": FA62233, "application-index": 27364, "items": [ a bunch of strings] },
  .
  .
]
```
And I'm not sure if you want one of these objects for _every_ top-level transaction, or only if the top-level txn is an app call.

Regardless, I think you need hierarchy, in order to convey that some of those logs were written by inner apps with different app-ids.  This might work.
```
[ {txid: "FA56E" logs: [{app-id: 123, str1}, {app-id:233: str2}...]},
   {txid: "A6B5E" logs: [{app-id: 311, str1}, {app-id:111: str2}...]},
]
```
Here a top-level app called an inner app, and each logged one thing.

joe-p replied:
I think the most common use case of this endpoint will be for parsing events. For events, you typically don't care about how it was called. Any necessary information will be encoded in the event itself. Having a nested structure will make it a bit harder to parse all the logs and see if an event you care about happened.

If we want to include "location" in the response then I would propose we add a `LoggedAt` field that resembles `FailedAt` from the simulate response:

> FailedAt If present, indicates which transaction in this group caused the failure. This array represents the path to the failing transaction. Indexes are zero based, the first element indicates the top-level transaction, and successive elements indicate deeper inner transactions.

jasonpaulos replied:
I like the current schema. It matches the "proposed solution" from #5807 which I helped define and was happy with.

My opinion on edge cases are:
* As @jannotti pointed out, txns with no logs should be filtered out (right now I think they're still present with an empty logs array)
* If a top-level app invokes the same app multiple times as an inner, I think the inners ought to be separate objects. E.g.
```json
[
  { "txid": "TOPLVL", "application-index": 123, "items": ["..."] },
  { "txid": "TOPLVL", "application-index": 123, "items": ["..."] }
]
```

Reasoning for the second bullet is that this is essentially the same as two top-level app calls to the same app in the same group, and in that case they would be separate objects. Plus it seems "nice" that each object represents a single transaction.

jasonpaulos replied:
> I think that you'll naturally get your second bullet point because each inner is processed independently right? There's no collection by id going on.

With the current code yes that naturally falls out. For a moment I was questioning what makes "the most sense" though, so I was considering alternatives.

jasonpaulos replied:
And to @joe-p's point, it would be easy to add a `FailedAt`-like field to point to exactly which inner was responsible for those logs. But in my opinion is that's not necessary based on our current expectations of how events will be processed. Can always revisit if there's a need for it though.

jasonpaulos commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,54 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
+		*blockLogs = append(*blockLogs, model.BlockLog{
+			Txid:             outerTxnID,
+			Logs:             byteLogs,
+			ApplicationIndex: uint64(txn.ApplicationID),
```
Two points I want to make here:
* Go embedding trickery: you might think this resolves to `txn.SignedTxn.Txn.ApplicationID`, but it's actually resolves to `txn.ApplyData.ApplicationID`. The `ApplicationID` name is duplicated across two different embedded structs, and for some reason (maybe definition order) go decides to resolve the `ApplyData` one. Best to be explicit in this case.
* In case this is an app create call (i.e. `txn.SignedTxn.Txn.ApplicationID == 0`), you _do_ want to actually reference `txn.ApplyData.ApplicationID`, as that will have the created app ID.

jasonpaulos replied:
Yes

joe-p replied:
Resolved in [0d8e597](https://github.com/algorand/go-algorand/pull/5865/commits/0d8e597107444f3e8218246e77b23ee5572ecccf)

joe-p commented:
I've verified the logs are returned properly via external testing here: https://github.com/joe-p/block-log-debug/blob/6daea7ad53b83366c75aaf3325c719bc6fbd0646/__test__/block-log-debug.test.ts#L12

I'll need to write the proper tests here but should be good to go. 

Other thing to consider:

1. Do we want query parameters to filter logs? I can imagine a prefix and appID filter would be useful. 
2. I feel like we should probably have some sort of way to limit response size. I think the most reasonable approach is to add query parameters for `start` and `maxResults` parameter but this will only work if txn order is deterministic in the block. I would think it is but would like some confirmation.

jannotti commented:
Since this returns, at most, the logs in one block, I don't think you need to add any filtering or limiting.

joe-p commented:
Yeah true, we know the upper bound due to the block size limit. Keeping it as is is good with me then.

jannotti commented on daemon/algod/api/algod.oas2.json:
```diff
@@ -787,6 +787,62 @@
+            }
+          },
+          "404": {
+            "description": "None existing block ",
```
```suggestion
            "description": "Nonexistent block ",
```

jannotti commented on daemon/algod/api/algod.oas2.json:
```diff
@@ -4441,6 +4497,33 @@
         }
       }
+    },
+    "BlockLog": {
```
Now that I understand the schema better (sorry about that), I think this object should not have `Block` in the name.  These are all of the logs for a single app-call, right? How about `AppCallLogs`?

jannotti commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,81 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
 	return ctx.JSON(http.StatusOK, response)
 }
 
+func NewBlockLog(txid string, logs []string, appIndex uint64) *model.BlockLog {
+	byteLogs := make([][]byte, len(logs))
+	for i, log := range logs {
+		byteLogs[i] = []byte(log)
+	}
+
+	return &model.BlockLog{
+		Txid:             txid,
+		Logs:             byteLogs,
+		ApplicationIndex: appIndex,
+	}
+}
```
We have a `convertSlice` in this package you can use.
```suggestion
func NewBlockLog(txid string, logs []string, appIndex uint64) *model.BlockLog {
	return &model.BlockLog{
		Txid:             txid,
		Logs:             convertSlice(logs, fn (s string) { return []byte(s) })
		ApplicationIndex: appIndex,
	}
}
```

jannotti commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,81 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
 	return ctx.JSON(http.StatusOK, response)
 }
 
+func NewBlockLog(txid string, logs []string, appIndex uint64) *model.BlockLog {
```
You can return `model.BlockLog` here. All call sites dereference and throw it away anyway.

jannotti reviewed (commented):
A few small things you may or may not bother with.

joe-p commented:
@jasonpaulos @jannotti I just added a fairly basic test in `test/e2e-go/restAPI/other/appsRestAPI_test.go`. Seemed like the most appropriate place for this to go but let me know what else is needed before this can be merged.

jasonpaulos commented on test/e2e-go/restAPI/other/appsRestAPI_test.go:
```diff
@@ -549,3 +549,85 @@ end:
+	txid := confirmation.Txn.ID()
+
+	// get app ID
+	submittedAppCreateTxn, err := testClient.PendingTransactionInformation(appCreateTxID)
+	a.NoError(err)
+	a.NotNil(submittedAppCreateTxn.ApplicationIndex)
+	createdAppID := basics.AppIndex(*submittedAppCreateTxn.ApplicationIndex)
+	a.NotZero(createdAppID)
```
minor: you don't need this section of code. `confirmation.ApplicationIndex` will give you the app ID

jasonpaulos commented on test/e2e-go/restAPI/other/appsRestAPI_test.go:
```diff
@@ -549,3 +549,85 @@ end:
+	a.NoError(err)
+	a.Equal(logs.Logs[0], deadBeef)
+
+	// delete the app
```
There's no need to delete the app. The code after this line can be removed

jasonpaulos commented on test/e2e-go/restAPI/other/appsRestAPI_test.go:
```diff
@@ -549,3 +549,85 @@ end:
+	a.NotZero(createdAppID)
+
+	// get block logs
+	resp, err := testClient.BlockLogs(*round)
```
I strongly suggest testing that you can get logs from an inner txn, since that's an essential part of this endpoint

jasonpaulos commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,76 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
+		if len(txn.EvalDelta.Logs) > 0 {
+			blockLogs = append(
+				blockLogs,
+				NewBlockLog(txn.ID().String(), txn.EvalDelta.Logs, getAppIndexFromTxn(txn)),
```
nit: it'd be nice to do something like `txID := txn.ID().String()` for eacn txn so you can avoid recalculating the txID

jasonpaulos commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,76 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
+			)
+		}
+
+		getLogsFromTxns(txn.EvalDelta.InnerTxns, blockLogs, outerTxnID)
```
This needs a `blockLogs = ` in front of it

joe-p commented:
Addressed @jasonpaulos comments and updated the tests to cover the following scenarios 

* log in outer transaction
* log from inner transaction
* log from newly created application
* logs from multiple outer app calls

jasonpaulos commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,77 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
 	return ctx.JSON(http.StatusOK, response)
 }
 
+func NewBlockLog(txid string, logs []string, appIndex uint64) model.AppCallLogs {
```
Review dog is right, this needs a comment before it

jasonpaulos commented on test/e2e-go/restAPI/other/appsRestAPI_test.go:
```diff
@@ -549,3 +550,144 @@ end:
+	deadBeef, err := hex.DecodeString("deadbeef")
+
+	// get block logs
+	resp, err := testClient.BlockLogs(*round)
+	a.NoError(err)
+	l0 := resp.Logs[0]
+	l1 := resp.Logs[1]
+	l2 := resp.Logs[2]
+	l3 := resp.Logs[3]
+
+	a.Equal(uint64(createdAppID), l0.ApplicationIndex)
+	a.Equal(stxn0.ID().String(), l0.Txid)
+	a.Equal(deadDood, l0.Logs[0])
+
+	a.Equal(uint64(createdAppID)+3, l1.ApplicationIndex)
+	a.Equal(stxn0.ID().String(), l1.Txid)
+	a.Equal(deadBeef, l1.Logs[0])
+
+	a.Equal(uint64(createdAppID), l2.ApplicationIndex)
+	a.Equal(stxn1.ID().String(), l2.Txid)
+	a.Equal(deadDood, l2.Logs[0])
+
+	a.Equal(uint64(createdAppID)+5, l3.ApplicationIndex)
+	a.Equal(stxn1.ID().String(), l3.Txid)
+	a.Equal(deadBeef, l3.Logs[0])
+}
```
I suggest this change in assertion style because it's much stronger -- if any additional fields/elements are present that shouldn't be, this will catch that.
```suggestion
    resp, err := testClient.BlockLogs(*round)
	a.NoError(err)

	expected := model.BlockLogsResponse{
		Logs: []model.AppCallLogs{
			{
				ApplicationIndex: uint64(createdAppID),
				Txid:             stxn0.ID().String(),
				Logs:             [][]byte{deadDood},
			},
			{
				ApplicationIndex: uint64(createdAppID + 3),
				Txid:             stxn0.ID().String(),
				Logs:             [][]byte{deadBeef},
			},
			{
				ApplicationIndex: uint64(createdAppID),
				Txid:             stxn1.ID().String(),
				Logs:             [][]byte{deadDood},
			},
			{
				ApplicationIndex: uint64(createdAppID + 5),
				Txid:             stxn1.ID().String(),
				Logs:             [][]byte{deadBeef},
			},
		},
	}
	a.Equal(expected, resp)
```

jasonpaulos commented on test/e2e-go/restAPI/other/appsRestAPI_test.go:
```diff
@@ -549,3 +550,144 @@ end:
+	a.NoError(err)
+
+	// call app twice
+	appCallTxn, err := testClient.MakeUnsignedAppNoOpTx(
```
missing error check

jasonpaulos commented on test/e2e-go/restAPI/other/appsRestAPI_test.go:
```diff
@@ -549,3 +550,144 @@ end:
+		uint64(createdAppID), nil, nil, nil,
+		nil, nil,
+	)
+	appCallTxn0, err := testClient.FillUnsignedTxTemplate(someAddress, 0, 0, 0, appCallTxn)
```
missing error check

jasonpaulos commented on test/e2e-go/restAPI/other/appsRestAPI_test.go:
```diff
@@ -549,3 +550,144 @@ end:
+
+	round := callConf.ConfirmedRound
+
+	deadDood, err := hex.DecodeString("deadd00d")
+	deadBeef, err := hex.DecodeString("deadbeef")
```
missing error check after each of these. There's no reason these should fail, but it's good practice (my linter complains locally, but I'm surprised reviewdog isn't also complaining)

jasonpaulos reviewed (commented):
A few comments. Willing to approve after those changes are made

jannotti commented on test/e2e-go/restAPI/other/appsRestAPI_test.go:
```diff
@@ -549,3 +550,154 @@ end:
+
+	round := callConf.ConfirmedRound
+
+	deadDood, err := hex.DecodeString("deadd00d")
+	a.NoError(err)
+	deadBeef, err := hex.DecodeString("deadbeef")
+	a.NoError(err)
```
Can you add a case for an empty log? (I think that's legal!)
And a case with an embedded null byte? I thought`deadd00d` had one at first glance, but the 00 is spread across bytes.
And a case with multiple logs from same app?
And a case with no logs? (To show we don't get an empty `AppCallLogs`)

joe-p replied:
Done in 733567f9038d9e14bfb76258607caf33209391dd

joe-p commented on daemon/algod/api/server/v2/handlers.go:
```diff
@@ -732,6 +732,78 @@ func (v2 *Handlers) GetBlockTxids(ctx echo.Context, round uint64) error {
+				blockLogs,
+				NewAppCallLogs(txid, txn.EvalDelta.Logs, getAppIndexFromTxn(txn)),
+			)
+		}
```
In [6209b64](https://github.com/algorand/go-algorand/pull/5865/commits/6209b64df57b55d7d3dfba849027d8a5e83e1327) changed it to just call the function with a slice of the outer + inners

joe-p commented on daemon/algod/api/algod.oas2.json:
```diff
@@ -4449,6 +4505,33 @@
         }
       }
+    },
+    "AppCallLogs": {
```
Let me know what you think of 340ad0a2e256dce88c04eaccd78422d1fd1bda5b

jannotti commented on daemon/algod/api/algod.oas2.json:
```diff
@@ -5118,6 +5201,23 @@
       }
     },
+    "BlockLogsResponse": {
+      "description": "All of the logs from outer and inner app calls in the given round. Each app call results in a new AppCallLogs object, thus there might be multiple objects with the same application ID and outer transaction ID in the event of multiple inner app calls to the same app. App calls with no logs are not included in the response.",
```
```suggestion
      "description": "All logs emitted in the given round. Each app call, whether top-level or inner, that contains logs results in a separate AppCallLogs object. Therefore there may be multiple AppCallLogs with the same application ID and outer transaction ID in the event of multiple inner app calls to the same app. App calls with no logs are not included in the response. AppCallLogs are returned in the same order that their corresponding app call appeared in the block (pre-order traversal of inner app calls)",
```

jannotti reviewed (approved):
I'm happy if we get some form of the description change I recommended. I want the response order to be defined.

jasonpaulos reviewed (commented):
+1 to being more explicit about the endpoint description, but other than that, this looks good

gmalouf commented:
closes #5807
